Fix #212: Unassign author if changelog label added

SAEb-ai opened the pull request:
<!--
  - Thanks for submitting code to Oppiabot! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fix #212.It unassigns the author if changelog label added by the author and if there are no review comments left.
## Checklist
- [X] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [X] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

jameesjohn commented on lib/checkPullRequestLabels.js:
```diff
@@ -210,6 +213,26 @@ module.exports.checkChangelogLabel = async function (context) {
     }
+    else if (!hasChangesBeingRequested) {
+      const author = pullRequest.user.login;
+      const allAssignees = pullRequest.assignees.map((assignee) => assignee.login);
+      if (allAssignees.includes(author)) {
```
This can be refactored to check if the author is assigned in just one iteration
```
const authorIsAssigned = pullRequest.assignees.some(assignee => assignee.login === author)
if (authorIsAssigned) {
```

SAEb-ai replied:
Done.

jameesjohn reviewed (changes requested):
Thanks, @SAEb-ai. I left some comments.
Also, the tests need to be updated to reflect this change.

SAEb-ai commented:
@jameesjohn PTAL.
I have addressed your review comments. I think the frontend tests are not providing 100% file coverage due to which its failing. 
I am unable to run the tests locally. Can you please help me in that too @jameesjohn.
![failfrontend](https://user-images.githubusercontent.com/69013647/111911380-64bb0f00-8a8b-11eb-9da9-62c9ed044407.JPG)

@vojtechjelinek PTAL at the PR.
Thanks!

vojtechjelinek commented on lib/checkPullRequestLabels.js:
```diff
@@ -210,6 +213,26 @@ module.exports.checkChangelogLabel = async function (context) {
+        await context.github.issues.createComment(
+          context.repo({
+            issue_number: pullRequest.number,
+            body: 'Unassigning @' + author + ' since the changelog label is added.',
```
```suggestion
            body: 'Unassigning @' + author + ' since the changelog label was added.',
```

SAEb-ai replied:
Done

vojtechjelinek commented on spec/checkPullRequestLabelsSpec.js:
```diff
@@ -793,9 +793,12 @@ describe('Pull Request Label Check', () => {
           color: '00FF00',
         };
 
+        const assigned = [{ login: 'SAEb-ai' }];
```
Could we use a more generic login here? So that it is not linked to your username?

SAEb-ai replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

SAEb-ai commented:
@jameesjohn I was now able to resolve my problem.(https://github.com/oppia/oppiabot/pull/216#issuecomment-803608599).

Here's the frontend coverage.
![f1](https://user-images.githubusercontent.com/69013647/111912636-8ec30000-8a90-11eb-905a-8eb2167afcc9.JPG)

Here's the error message.
![F2](https://user-images.githubusercontent.com/69013647/111912687-bd40db00-8a90-11eb-97bd-b823839f1ee6.JPG)

@jameesjohn Can you please help me to resolve this.

SAEb-ai commented:
@vojtechjelinek  PTAL.
I have addressed your review comments.

SAEb-ai commented on lib/checkPullRequestLabels.js:
```diff
@@ -210,6 +213,26 @@ module.exports.checkChangelogLabel = async function (context) {
+      const author = pullRequest.user.login;
+      const authorIsAssigned = pullRequest.assignees.some(assignee => assignee.login === author)
+      if (authorIsAssigned) {
+        //Unassign the author as the changelog label is added
```
Done

jameesjohn commented on lib/checkPullRequestLabels.js:
```diff
@@ -210,6 +213,26 @@ module.exports.checkChangelogLabel = async function (context) {
     }
+    else if (!hasChangesBeingRequested) {
+      const author = pullRequest.user.login;
+      const authorIsAssigned = pullRequest.assignees.some(assignee => assignee.login === author)
```
The above line is longer than 80 characters.

jameesjohn reviewed (changes requested):
Thanks, I've left some comments.

The tests are failing because you'd need to mock `context.github.search.issuesAndPullRequests` which `utils.doesPullRequestHaveChangesRequested` uses under the hood.

You should create a new describe block to test the new feature, and then you can mock the above function.

SAEb-ai commented:
@jameesjohn I just want to ask you whether I am going in the right way to mock `context.github.search.issuesAndPullRequests` and then I will proceed further.

jameesjohn commented:
@SAEb-ai, you can check out https://github.com/oppia/oppiabot/blob/862db8dc81072dff03b239ffe0cd730bc4d39b8e/spec/checkBranchPushSpec.js#L48-L55

jameesjohn commented:
Also, when responding to reviews, do well to reply to each comment so I can know which has been attended to.
Thanks.

SAEb-ai commented on spec/checkPullRequestLabelsSpec.js:
```diff
@@ -793,9 +793,12 @@ describe('Pull Request Label Check', () => {
+        payloadData.payload.pull_request.user.login = 'username';
         payloadData.payload.pull_request.labels = [label];
+        payloadData.payload.pull_request.assignees = assigned;
+        payloadData.payload.pull_request.review_comments = 0;
```
Done

SAEb-ai commented:
@jameesjohn I have made a new describe block (https://github.com/oppia/oppiabot/pull/216#discussion_r598615005) as requested by you. I am only stuck at whether my mocking procedure is correct or not.
Thanks!

vojtechjelinek reviewed (changes requested):
Thanks! Only one small nit.

SAEb-ai commented on constants.js:
```diff
@@ -100,7 +100,7 @@ const checksWhitelist = {
     [issueCommentCreatedEvent]: [respondToReviewCheck]
   },
-  'oppiabot': {
+  oppiabot: {
```
Sorry for it. I don't how it came but I changed it to the original version in my recent commit. Thanks!

SAEb-ai commented:
@jameesjohn I have now mocked `context.github.search.issuesAndPullRequests` but now the following error is coming.PTAL.
```
Failures:
1) Pull Request Label Check when pull request gets opened or reopened adds a default label when pr author is not a collaborator and PR reviewers are more than one
  Message:
    Error: Timeout - Async function did not complete within 5000ms (set by jasmine.DEFAULT_TIMEOUT_INTERVAL)
  Stack:
    Error: Timeout - Async function did not complete within 5000ms (set by jasmine.DEFAULT_TIMEOUT_INTERVAL)
        at <Jasmine>
        at listOnTimeout (node:internal/timers:557:17)
        at processTimers (node:internal/timers:500:7)

2) Pull Request Label Check when pull request gets opened or reopened should not ping pr author if there is a changelog label
  Message:
    Error: Timeout - Async function did not complete within 5000ms (set by jasmine.DEFAULT_TIMEOUT_INTERVAL)
  Stack:
    Error: Timeout - Async function did not complete within 5000ms (set by jasmine.DEFAULT_TIMEOUT_INTERVAL)
        at <Jasmine>
        at listOnTimeout (node:internal/timers:557:17)
        at processTimers (node:internal/timers:500:7)

3) Pull Request Label Check when valid changelog added with no review comments should unassign the author
  Message:
    Expected spy doesPullRequestHaveChangesRequested to have been called.
  Stack:
    Error: Expected spy doesPullRequestHaveChangesRequested to have been called.
        at <Jasmine>
        at UserContext.<anonymous> (/home/shahbaz/opensource/oppiabot/spec/checkPullRequestLabelsSpec.js:37:2546)
        at processTicksAndRejections (node:internal/process/task_queues:94:5)
  Message:
    Error: <toHaveBeenCalled> : Expected a spy, but got undefined.
    Usage: expect(<spyObj>).toHaveBeenCalled()
  Stack:
    Error: <toHaveBeenCalled> : Expected a spy, but got undefined.
    Usage: expect(<spyObj>).toHaveBeenCalled()
        at <Jasmine>
        at UserContext.<anonymous> (/home/shahbaz/opensource/oppiabot/spec/checkPullRequestLabelsSpec.js:37:3102)
        at processTicksAndRejections (node:internal/process/task_queues:94:5)

495 specs, 3 failures
Finished in 97.622 seconds
```

SAEb-ai commented:
@vojtechjelinek Thanks for review comments. I have addressed them.PTAL.

vojtechjelinek reviewed (approved):
LGTM!

SAEb-ai commented:
@vojtechjelinek I have one more doubt in it. Consider if the  contributor mistakenly unassigns
him even before adding the Changelog level, should oppiabot not ping the author that he should not unassign himself before adding the change log level and assign the contributor again. Does it make sense?
Thanks!

SAEb-ai commented:
@jameesjohn @vojtechjelinek Could please reply to  https://github.com/oppia/oppiabot/pull/216#issuecomment-805794362   so that I can update  the new changes to this branch if it sounds good.
Thanks!

vojtechjelinek commented:
> @vojtechjelinek I have one more doubt in it. Consider if the contributor mistakenly unassigns
> him even before adding the Changelog level, should oppiabot not ping the author that he should not unassign himself before adding the change log level and assign the contributor again. Does it make sense?
> Thanks!

I cannot really understand what are you asking, could you please try to rephrase it?

SAEb-ai commented:
@vojtechjelinek. Consider :
1. A PR is made without a PR Changelog Label.
2. Oppiabot assigns the author(collaborator) for adding the changelog level.
3. The collaborator mistakenely unassigns himself before adding the PR changelog level.

My question was after the third step should oppiabot not warn the author that he should first add the changelog level and then he can unassign himself.
It simply means making oppiabot behave strictly towards new contributors.
Hope it makes sense. If it doesn't, please ping me.
Thanks!

vojtechjelinek commented:
> @vojtechjelinek. Consider :
> 
>     1. A PR is made without a PR Changelog Label.
> 
>     2. Oppiabot assigns the author(collaborator) for adding the changelog level.
> 
>     3. The collaborator mistakenely unassigns himself before adding the PR changelog level.
> 
> 
> My question was after the third step should oppiabot not warn the author that he should first add the changelog level and then he can unassign himself.
> It simply means making oppiabot behave strictly towards new contributors.
> Hope it makes sense. If it doesn't, please ping me.
> Thanks!
I mean you can add it. Btw, do we handle the case when the author cannot add the label so it doesn't make sense to assign them?

SAEb-ai commented:
> > @vojtechjelinek. Consider :
> > ```
> > 1. A PR is made without a PR Changelog Label.
> > 
> > 2. Oppiabot assigns the author(collaborator) for adding the changelog level.
> > 
> > 3. The collaborator mistakenely unassigns himself before adding the PR changelog level.
> > ```
> > 
> > 
> > My question was after the third step should oppiabot not warn the author that he should first add the changelog level and then he can unassign himself.
> > It simply means making oppiabot behave strictly towards new contributors.
> > Hope it makes sense. If it doesn't, please ping me.
> > Thanks!
> > I mean you can add it. Btw, do we handle the case when the author cannot add the label so it doesn't make sense to assign them?

@vojtechjelinek I mean to say that this is meant only for collaborators

SAEb-ai commented:
@jameesjohn Could you please reply to my https://github.com/oppia/oppiabot/pull/216#issuecomment-805505139 
Thanks!

jameesjohn commented:
> @jameesjohn Could you please reply to my #216 (comment)
Thanks!

Sorry for the late response @SAEb-ai.

To fix the timeout issue with the tests, change 
https://github.com/oppia/oppiabot/blob/05941afe45ea7c129cbbd81852bfdeabe0f5ebd6/lib/checkPullRequestLabels.js#L220
to
`await utilityModule.sleep(5000);`

and add `spyOn(utilityModule, 'sleep').and.callFake(() => { });` to https://github.com/oppia/oppiabot/blob/05941afe45ea7c129cbbd81852bfdeabe0f5ebd6/spec/checkPullRequestLabelsSpec.js#L48-L49

vojtechjelinek commented on constants.js:
```diff
@@ -100,7 +100,7 @@ const checksWhitelist = {
     [issueCommentCreatedEvent]: [respondToReviewCheck]
   },
-  'oppiabot': {
+'oppiabot': {
```
Unneeded change.

SAEb-ai replied:
TODO:After failure of frontend tests get fixed.

SAEb-ai commented on spec/checkPullRequestLabelsSpec.js:
```diff
@@ -766,7 +780,7 @@ describe('Pull Request Label Check', () => {
       ).and.callThrough();
       await robot.receive(payloadData);
-
+      
```
TODO:After Failure of frontend tests get fixed.

SAEb-ai commented on spec/checkPullRequestLabelsSpec.js:
```diff
@@ -743,7 +757,7 @@ describe('Pull Request Label Check', () => {
       ).and.callThrough();
       await robot.receive(payloadData);
-
+      
```
TODO:After Failure of frontend tests get fixed.

SAEb-ai commented on spec/checkPullRequestLabelsSpec.js:
```diff
@@ -693,7 +707,7 @@ describe('Pull Request Label Check', () => {
 
         await robot.receive(payloadData);
-
+        
```
TODO:After Failure of frontend tests get fixed.

SAEb-ai replied:
TODO:After Failure of frontend tests get fixed.

SAEb-ai commented on spec/checkPullRequestLabelsSpec.js:
```diff
@@ -630,7 +644,7 @@ describe('Pull Request Label Check', () => {
 
         await robot.receive(payloadData);
-
+        
```
TODO:After Failure of frontend tests get fixed.

SAEb-ai commented:
@jameesjohn Now there are not any timeout related failures for the frontend.The only error that is coming is
```
 Pull Request Label Check when valid changelog added with no review comments should unassign the author
  Message:
    Expected spy doesPullRequestHaveChangesRequested to have been called.
  Stack:
    Error: Expected spy doesPullRequestHaveChangesRequested to have been called.
        at <Jasmine>
        at UserContext.<anonymous> (/home/shahbaz/opensource/oppiabot/spec/checkPullRequestLabelsSpec.js:37:978)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
  Message:
    Error: <toHaveBeenCalled> : Expected a spy, but got undefined.
    Usage: expect(<spyObj>).toHaveBeenCalled()
  Stack:
    Error: <toHaveBeenCalled> : Expected a spy, but got undefined.
    Usage: expect(<spyObj>).toHaveBeenCalled()
        at <Jasmine>
        at UserContext.<anonymous> (/home/shahbaz/opensource/oppiabot/spec/checkPullRequestLabelsSpec.js:37:1534)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)


```
I have tried a lot of things and also updated those in my recent commits. But still the above frontend test failure is coming.Am I missing on  something?
PTAL.

Thanks!

jameesjohn commented:
> Message:
    Expected spy doesPullRequestHaveChangesRequested to have been called.

This is happening because of the way `doesPullRequestHaveChangesRequested` got imported into the pullRequestLabelsModule.
Update line 200 to `const hasChangesRequested = utilityModule.doesPullRequestHaveChangesRequested(`

SAEb-ai commented:
@jameesjohn Now I am getting this error.
![gp](https://user-images.githubusercontent.com/69013647/114381198-7bdfaf00-9ba8-11eb-816f-ff753ed23af8.JPG)

jameesjohn commented on spec/checkPullRequestLabelsSpec.js:
```diff
@@ -782,29 +787,71 @@ describe('Pull Request Label Check', () => {
+      };
+      payloadData.payload.action = 'reopened';
+      // Add changelog label.
+      payloadData.payload.pull_request.labels.push(label);
```
Can you set this to be the only label?

jameesjohn commented on spec/checkPullRequestLabelsSpec.js:
```diff
@@ -782,29 +787,71 @@ describe('Pull Request Label Check', () => {
+          .createSpy('issuesAndPullRequests')
+          .and.resolveTo({
+            data: {
+              items: [payloadData.payload.pull_request],
```
We want to check this when the pull request does not have changes requested. So this should be an empty array. 
You can also have a block that checks that this does not unassign the PR author when changes are requested.

jameesjohn commented on spec/checkPullRequestLabelsSpec.js:
```diff
@@ -782,29 +787,71 @@ describe('Pull Request Label Check', () => {
+      const params = {
+        repo: payloadData.payload.repository.name,
+        owner: payloadData.payload.repository.owner.login,
+        number: payloadData.payload.number,
```
```suggestion
        issue_number: payloadData.payload.number,
```

jameesjohn reviewed (changes requested):
Hi @SAEb-ai,  Sorry for the delay in review. 
Following the suggestions should help in the tests..
